code-demos/code_demo_logreg.Rmd

```diff
@@ -0,0 +1,431 @@
+i.e., we are modeling the posterior probabilities $\pi(x)$ of the labels
+directly using the logistic function, s.t.
+
+$$\pi(x) = \mathbb{P}(y=1 |x) = \frac{1}{1+\exp(-\theta^Tx)}.$$
+**Note** As in the lecture we suppress the intercept in notation, i.e., 
+$\theta^Tx \equiv \theta_0 + \theta^Tx$.
+
+This can be easily implemented as:
+```{r logreg-logistic}
+logistic <- function(X, theta, Xtheta= X%*% theta){
+  1  / (1+ exp(-Xtheta))
+}
+```
+
+And we can observe that logistic regression “squashes” the estimated linear 
+scores $\theta^Tx$ to $[0, 1]$:
+
+```{r logreg-plot_logistic}
+num_data <- 100
+theta <- c(0, 1)
+
+ggplot(data=data.frame(x = seq(-5,5,length.out = num_data)), aes(x)) +
+ stat_function(fun = function(x) logistic(cbind(rep(1,length(x)),x),theta))
+```
+
+As shown in the lecture from the definition of accuracy and the encoding of the 
+labels to $\{-1, 1\}$ one can derive the Bernoulli loss:
+$$L(y,f(x)) = \log(1 + \exp(-yf(x)))$$
+For the logistic regression it holds, that $f(x) = \theta^Tx$.
+```{r logreg-ber_loss}
+loss_bernoulli <- function(Y, X, theta, exp_yxt = exp(Y*(X%*%theta))){
+  log(1+1/exp_yxt)
+}
+```
+We can plot the loss for $y = 1$ and $y = -1$:
+```{r logreg-ber_loss_plot}
+library(reshape2)
+num_data <- 100
+theta <- c(0, 1)
+y <- c(1, -1)
+
+x =  seq(-4,4,length.out = num_data)
+L = sapply(y, function(y) loss_bernoulli(y,cbind(rep(1,length(x)),x),theta))
+plot_data = melt(data.frame(x = x, L1 = L[,1], L2 = L[,2]),id.vars = "x")
+
+ggplot(plot_data, aes(x = x, y=value, colour = variable)) + 
+  geom_line() +
+ labs(x=expression(f(x)), y=expression(L(y,f(x)))) +
+      scale_color_manual(name="", values=c("purple","yellow"), labels=c("y = 1", "y = -1")) 
+
+
+```
+
+
+
+Which means in order to minimize the loss, we should predict 
+$y = 1$ if 
+$$\theta^Tx \geq 0 \iff \pi(x) = \mathbb{P}(y=1 |x) = \frac{1}{1+\exp(-\theta^Tx)} \geq 0.5.$$ 
+```{r logreg-classify}
+classify_logreg <- function(X, theta, Xtheta = X %*% theta, conv = as.numeric){
+  probability <- logistic(X, theta,  Xtheta = Xtheta)
+  list(prediction = conv(ifelse(probability > .5, 1, -1)), 
+       levels = c("1","-1"))
+}
+```
+
+With the help of the loss function, we can get the empirical risk as usual:
+$$R_{emp} = \sum_{i=1}^nL(y^{(i)}, f(x^{(i)})$$
+
+```{r logreg-ber_risk}
+risk_bernoulli <- function(Y, X, theta, exp_yxt = exp(Y*(X%*%theta))){
+  sum(loss_bernoulli(Y, X, theta, exp_yxt))
+}
+```
+
+## Logistic regression - example
+
+To investigate how the parameter $\theta$ of
+the logistic regression can be estimated, we create some artificial data:
+
+```{r logreg-dgp}
+set.seed(1337)
+
+n <- 1000
+p <- 2
+
+# model without intercept
+X <- matrix(runif(n*p, -5, 5), nrow = n)
+theta <- c(3, -1)
+
+Y <- 2 * rbinom(n = n, size = 1, prob = logistic(X, theta)) - 1
+```
+
+We use the *plot_2D_classify* function from the [KNN code demo](./code_demo_knn.pdf) as we want to plot how well we 
```

Comment:
not sure that link is going to work like this later. plz open an issue for this to make sure we get the cross-refs right once thje new structure of the course website etc is fixed.